### PR TITLE
🐛  Consistent ordering for deletion priority

### LINKED ISSUE #6299: Scaling down a MachineDeployment with Machines with the same delete priority may result in too many machines being deleted
**What steps did you take and what happened:**
- Deploy a cluster and set the DeletePolicy for all MachineDeployments to `Oldest` or `Newest`.
- Scale any MachineDeployment up by >= 2.
- As new machines are provisioning, scale down the MachineDeployment by 1.
Since all of the currently provisioning machines have the same delete priority as any machine with a non-zero deletion timestamp, this can result in multiple machines (all of the currently provisioning machines) to be deleted.

**What did you expect to happen:**
Exactly one machine should be deleted when scaling down a MachineDeployment by 1.

**Environment:**

- Cluster-api version: 1.0.2 
- Kubernetes version: (use `kubectl version`): Tested with 1.21.9, 1.21.10, 1.22.3
- OS (e.g. from `/etc/os-release`): Any

/kind bug
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]
/area machine

### PR DESCRIPTION
**Pull request by thedadams**
If two machines had the same deletion priority, then on subsequent syncReplicas
calls different machines could be deleted. For example, when using Oldest or
Newest deletion priorities, then machines that are already deleting and machines
without a node ref have the same priority. Then this could lead to multiple
machines getting deleted on different calls to syncReplicas.

That is, if a MachineSet is scaled down by 1, then one machine would get
deleted, but then a second call to syncReplicas could delete a machine with no
node ref instead of the machine that is already deleting.

**What this PR does / why we need it**:
By ensuring consistent ordering for machines with the same deletion priority, it can be assured that the correct number of machines get deleted when scaling down a MachineDeployment.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #6299

### CODE REVIEW

**Comment by enxebre** [comment]
Thanks! Change seems reasonable to me.

>That is, if a MachineSet is scaled down by 1, then one machine would get
deleted, but then a second call to syncReplicas could delete a machine with no
node ref instead of the machine that is already deleting.

I'm curious have you seen this actually happening? i.e two different machines being returned in different calls in case of a tie? Change seems reasonable to me but regardless I'm assuming there's already a hidden determinism preventing this from happening?

/ok-to-test

**Comment by thedadams** [comment]
Yes, we have seen the behavior I described in the description: scale a deployment up by two, as machines are provisioning scale down by one, because missing node ref and deleting machines have the same priority (we use delete oldest by default) then both provisioning machines get deleted and another machine gets provisioned once they are gone.

**Review by killianmuldoon — APPROVED** [review]
/lgtm

**Comment by fabriziopandini** [comment]
/lgtm
@vincepri PTAL

**Comment by thedadams** [comment]
@fabriziopandini @vincepri Is there anything I should be doing for this PR to move forward?

**Comment by enxebre** [comment]
>I wonder a bit if we can fix this more holistically by treating machines already in deleting differently. But as this PR already fixes a known bug:

I'd be cautious about that since there're many nuances where we could impact by accident changing existing API behaviour. I agree this is a right bugfix with a reduced impact surface and further discussion can happen separately.

**Comment by sbueringer** [comment]
> > I wonder a bit if we can fix this more holistically by treating machines already in deleting differently. But as this PR already fixes a known bug:
> 
> I'd be cautious about that since there're many nuances where we could impact by accident changing existing API behaviour unexpectedly. I agree this is a right bugfix with a reduced impact surface and further discussion can happen separately.

Yeah me too. I looked at the surrounding code and that's why I definitely prefer "just" a bugfix for now.
I was just wondering in general if we could avoid deleting machines which already have deletionTimestamps (I know it's a no-op). In cases where you're currently "deleting" Machines are not selected for deletion again you might end up with less Machines than you want to. But I'm not sure if that's a problem that can really occur.

so tl;dr I definitely prefer just a bugfix for now.

File: internal/controllers/machineset/machineset_delete_policy.go [hunk 1]
```diff
@@ -97,7 +97,11 @@ type sortableMachines struct {
 func (m sortableMachines) Len() int      { return len(m.machines) }
 func (m sortableMachines) Swap(i, j int) { m.machines[i], m.machines[j] = m.machines[j], m.machines[i] }
 func (m sortableMachines) Less(i, j int) bool {
-	return m.priority(m.machines[j]) < m.priority(m.machines[i]) // high to low
+	priorityI, priorityJ := m.priority(m.machines[i]), m.priority(m.machines[j])
+	if priorityI == priorityJ {
+		return m.machines[i].Name < m.machines[j].Name
```
Comment (vincepri) [inline internal/controllers/machineset/machineset_delete_policy.go:102]: Add a comment above this line to explain why we're falling back to name?

Comment (thedadams) [reply internal/controllers/machineset/machineset_delete_policy.go:104]: I added a comment, as you suggested.
